Mathlib/CategoryTheory/RegularCategory/Basic.lean

```diff
@@ -0,0 +1,173 @@
+* <https://ncatlab.org/nlab/show/regular+category>
+-/
+
+open CategoryTheory Limits
+
+universe u v
+
+namespace CategoryTheory
+
+section
+
+variable (C : Type u) [Category.{v} C]
+
+/--
+A regular category is a category with finite limits, such that every kernel pair has a coequalizer,
+and such that regular epimorphisms are stable under base change.
+-/
+class Regular extends HasFiniteLimits C where
+  hasCoequalizer_of_isKernelPair {a b c : C} {f : a ⟶ b} {g1 g2 : c ⟶ a} :
+    IsKernelPair f g1 g2 → HasCoequalizer g1 g2
+  regularEpiIsStableUnderBaseChange : MorphismProperty.IsStableUnderBaseChange (.regularEpi C)
+
+end
+
+variable {C : Type u} [Category.{v} C] [Regular C]
+variable {X Y : C} (f : X ⟶ Y)
+
+instance : Preregular C where
+  exists_fac f g :=
+    let : RegularEpi (pullback.snd g f) := by
+      have := Regular.regularEpiIsStableUnderBaseChange.of_isPullback
+        (IsPullback.of_hasPullback g f) ⟨inferInstance⟩
+      simp at this
+      exact regularEpiOfIsRegularEpi <| pullback.snd g f
+    ⟨pullback g f, pullback.snd g f, inferInstance, pullback.fst g f, pullback.condition⟩
+
+noncomputable section
+namespace Regular.StrongEpiMonoFactorisation
+
+local instance : HasCoequalizer (pullback.fst f f) (pullback.snd f f) :=
+  Regular.hasCoequalizer_of_isKernelPair <| IsKernelPair.of_hasPullback f
+
+/--
+The midpoint of the chosen mono factorisation associated to any regular category.
+It is defined to be the coequalizer of the kernel pair of an arrow `f`.
+See `monoFactorisation` for the actual mono factorisation.
+-/
+def I : C :=
+  coequalizer (pullback.fst f f) (pullback.snd f f)
+
+/--
+The `e` component of the chosen mono factorisation associated to any regular category.
+It is defined to be the projection from the coequalizer of the kernel pair of an arrow `f`.
+See `monoFactorisation` for the actual mono factorisation.
+-/
+def e : X ⟶ (I f) :=
+  coequalizer.π (pullback.fst f f) (pullback.snd f f)
+
+/-- The canonical regular epi structore on `e f`, as the coequalizer of the kernel pair of `f`. -/
+local instance eRegularEpi : RegularEpi (e f) := by
+  dsimp [e]; infer_instance
+
+instance e_isRegularEpi : IsRegularEpi (e f) := by
+  dsimp [e]; infer_instance
+
+/--
+The `mono` component of the chosen mono factorisation associated to any regular category.
+It is induced by the coequalizer of the kernel pair of an arrow `f`.
+See `monoFactorisation` for the actual mono factorisation.
+-/
+def m : (I f) ⟶ Y :=
+  coequalizer.desc f (IsKernelPair.of_hasPullback f).w
+
+@[reassoc (attr := simp)]
+lemma fac : (e f) ≫ (m f) = f :=
+  coequalizer.π_desc f _
+
+instance m_mono : Mono (m f) := by
+  apply (IsKernelPair.of_hasPullback (m f)).mono_of_eq_fst_snd
+  let k₁ := pullback.fst (m f) (m f)
+  let k₂ := pullback.snd (m f) (m f)
+  let d : _ ⟶ (pullback (m f) (m f)) :=
+    pullback.lift (pullback.fst f f ≫ e f) (pullback.snd f f ≫ e f)
+      (by simp; rw [pullback.condition])
```

Comment:
```suggestion
      (by simp [pullback.condition])
```

Reply:
This was a nonterminal simp in your code; if my fix hadn't worked you should have squeezed that simp (i.e. replaced it with the output of `simp?`).